### DISCUSSION TITLE
Monolith face and mood-tracker app

### DISCUSSION THREAD

**Discussion by revpriest** [wasp-os/wasp-os#489] [Show and tell]
Thanks for this WaspOS stuff, it's brilliant. I've had a ball all month programming my watch. I did not expect to spend January doing that.

I have some pull-requests if you want them? Not sure if it's being maintained here much really.

1) Monolithic Watch Face + font-apps
2) A Mood-tracking app
3) A 3-line update to the "software" app to have it call any "unregister" function that exists on the app it's uninstalling.

The fonts and the mood-tracker register themselves by adding function to the wasp global, and then are used by the monolithic app if they are present. The unregister function allows them to free that memory as they are uninstalled.

The app and face are described in my image-laden blog post here:
https://dalliance.net/vPre2/waspos

In order to discover each other the mood and font apps just write to wasp.font and wasp.mood and the face looks for them there. I have no idea if this approach is good, or even works outside PineTime and the sim. That's why the un-register ended up being useful. If there's a proper registration system I could presumably reprogram to use that but I couldn't see anything.

Do we want these pull requests? I can send them in bulk or as three separate ones.

Here's my fork: https://github.com/revpriest/pre-wasp-os

My monolith app covers most of the functions of the default settings and steps and I'm tight on RAM so I commented them out. That exposed a bug in simulator/watch.py that resulted in a undefined var exception. I have a one-line fix for that too if we want it submitted.

Haven't really got much experience with using GitHub for merge requests and things. Dunno what the normal protocol is. 

![MonolithAppApp](https://github.com/wasp-os/wasp-os/assets/812357/f683c6f7-aa9b-4c58-a860-9ae8de00f3c8)

**Comment by daniel-thompson** [comment]
The basics are covered in the contributors guide (https://wasp-os.readthedocs.io/en/latest/contributing.html ). Which includes a couple of hints about how to make sure every commit has a DCO sign off. There's a bank of automated tests... but you can check the results of this using the web UI (once someone approves running them). They run a bunch of checks on the simulator and verify that the app has screenshots, etc, etc.

Wasp-os is currently looking for more mainainers but I am still merging stuff occasionally (there can sometimes be a bit of a wait) when pull requests come in.

FWIW I'd also suggest moving the date down a few pixels to get a bigger gap between the icon row. Personally I also prefer uppercase at the begining of labels but if lowercase is part of your aestetic choices I won't put up much of a fight. It sounds like a big app if it's nudging the limits of the RAM/flash so it would be good to mention that in the docs.

**Comment by revpriest** [comment]
Great. I've changed the label-case and the date-string position, good suggestions.

Might leave it a week or two before I make any pull-requests. Give a bit of time for the bugs to be shaken out. Fixed another one  this morning. Perhaps will wait until there's been a few days without one. ;)

**Comment by revpriest** [comment]
"Mention that in the docs", would that be in docs/apps.rst? Or just in the source-code for the app?

Is there anywhere I can learn about running these automated tests on my own machine or github fork?

**Reply by fgaz** [reply]
You can run `make check` and/or enable CI in the actions tab of your fork

**Comment by revpriest** [comment]
Oh, that's easy. Thanks.
There are indeed fails. All those font apps will need screenshots for a start I guess.
